Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,55 +49,162 @@ void RebinToWorkspace::init() {
-  setProperty("OutputWorkspace", ws);
+  m_preserveEvents = getProperty("PreserveEvents");
+  m_isEvents = bool(boost::dynamic_pointer_cast<const EventWorkspace>(toRebin));
+
+  if (needToRebin(toRebin, toMatch)) {
+    g_log.information("Rebinning");
+    if (m_isEvents && (!m_preserveEvents)) {
+      this->histogram(toRebin, toMatch);
+    } else {
+      this->rebin(toRebin, toMatch);
+    }
+  } else { // don't need to rebin
+    g_log.information(
+        "WorkspaceToRebin and WorkspaceToMatch already have matched binning");
+    auto outputWS = this->createOutputWorkspace(toRebin, 1.);
+    this->setProperty("OutputWorkspace", outputWS);
+  }
 }
 
-/**
- * Create the vector of rebin parameters
- * @param toMatch :: A shared pointer to the workspace with the desired binning
- * @returns :: A vector to hold the rebin parameters once they have been
- * calculated
- */
-std::vector<double> RebinToWorkspace::createRebinParameters(
-    Mantid::API::MatrixWorkspace_sptr toMatch) {
-  using namespace Mantid::API;
-
-  auto &matchXdata = toMatch->x(0);
-  // params vector should have the form [x_1, delta_1,x_2, ...
-  // ,x_n-1,delta_n-1,x_n), see Rebin.cpp
-  std::vector<double> rb_params;
-  int xsize = static_cast<int>(matchXdata.size());
-  rb_params.reserve(xsize * 2);
-  for (int i = 0; i < xsize; ++i) {
-    // bin bound
-    rb_params.push_back(matchXdata[i]);
-    // Bin width
-    if (i < xsize - 1)
-      rb_params.push_back(matchXdata[i + 1] - matchXdata[i]);
+MatrixWorkspace_sptr
+RebinToWorkspace::createOutputWorkspace(MatrixWorkspace_sptr &inputWS,
+                                        const double endProgress) {
+  MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+
+  bool inPlace = (inputWS == outputWS);
+
+  if (inPlace) {
+    if (m_isEvents && (!m_preserveEvents)) {
+      // convert to a MatrixWorkspace
+      IAlgorithm_sptr convert =
+          createChildAlgorithm("ConvertToMatrixWorkspace");
+      convert->setProperty<MatrixWorkspace_sptr>("InputWorkspace", inputWS);
+      convert->setProperty<MatrixWorkspace_sptr>("OutputWorkspace", outputWS);
+      convert->setChildStartProgress(0.);
+      convert->setChildStartProgress(endProgress);
+      convert->executeAsChildAlg();
+      outputWS = convert->getProperty("OutputWorkspace");
+    }
+    // all other cases are already handled by not doing anything
+  } else {
+    if (m_isEvents && (!m_preserveEvents)) {
+      // convert to a MatrixWorkspace
+      IAlgorithm_sptr convert =
+          createChildAlgorithm("ConvertToMatrixWorkspace");
+      convert->setProperty<MatrixWorkspace_sptr>("InputWorkspace", inputWS);
+      convert->setProperty<MatrixWorkspace_sptr>("OutputWorkspace", outputWS);
+      convert->setChildStartProgress(0.);
+      convert->setChildStartProgress(endProgress);
+      convert->executeAsChildAlg();
+      outputWS = convert->getProperty("OutputWorkspace");
+    } else {
+      outputWS = inputWS->clone();
+    }
```

Comment:
Can you swap the inner and outer `if` to avoid the large amount of code duplication? Like so:

```cpp
if (m_isEvents && (!m_preserveEvents)) {
  // convert to a MatrixWorkspace
  IAlgorithm_sptr convert = createChildAlgorithm("ConvertToMatrixWorkspace");
  convert->setProperty<MatrixWorkspace_sptr>("InputWorkspace", inputWS);
  convert->setProperty<MatrixWorkspace_sptr>("OutputWorkspace", outputWS);
  convert->setChildStartProgress(0.);
  convert->setChildStartProgress(endProgress);
  convert->executeAsChildAlg();
  outputWS = convert->getProperty("OutputWorkspace");
} else {
  if (!inPlace) {
    outputWS = inputWS->clone();
  } else {
    // all other cases are already handled by not doing anything
  }
}
```

If we go on reading, we can simplify more, see next comment.